packages/workbox-streams/concatenate.mjs

```diff
@@ -0,0 +1,147 @@
+ limitations under the License.
+*/
+
+import {logger} from 'workbox-core/_private/logger.mjs';
+import {assert} from 'workbox-core/_private/assert.mjs';
+
+import './_version.mjs';
+
+/**
+ * Takes either a Response, a ReadableStream, or a
+ * [BodyInit](https://fetch.spec.whatwg.org/#bodyinit) and returns the
+ * ReadableStreamReader object associated with it.
+ *
+ * @param {workbox.streams.StreamSource} source
+ * @return {ReadableStreamReader}
+ * @private
+ */
+function _getReaderFromSource(source) {
+  if (source.body && source.body.getReader) {
+    return source.body.getReader();
+  }
+
+  if (source.getReader) {
+    return source.getReader();
+  }
+
+  // TODO: This should be possible to do by constructing a ReadableStream, but
+  // I can't get it to work. As a hack, construct a new Response, and use the
+  // reader associated with its body.
+  return new Response(source).body.getReader();
+}
+
+/**
+ * Takes multiple source Promises, each of which could resolve to a Response, a
+ * ReadableStream, or a [BodyInit](https://fetch.spec.whatwg.org/#bodyinit).
+ *
+ * Returns an object exposing a ReadableStream with each individual stream's
+ * data returned in sequence, along with a Promise which signals when the
+ * stream is finished (useful for passing to a FetchEvent's waitUntil()).
+ *
+ * @param {Array<Promise<workbox.streams.StreamSource>>} sourcePromises
+ * @return {Object<{done: Promise, stream: ReadableStream}>}
+ *
+ * @memberof workbox.streams
+ */
+function concatenate(sourcePromises) {
+  if (process.env.NODE_ENV !== 'production') {
+    assert.isArray(sourcePromises, {
+      moduleName: 'workbox-streams',
+      funcName: 'concatenate',
+      paramName: 'sourcePromises',
+    });
+  }
+
+  const readerPromises = sourcePromises.map((sourcePromise) => {
+    return Promise.resolve(sourcePromise).then((source) => {
+      return _getReaderFromSource(source);
+    });
+  });
+
+  let fullyStreamedResolve;
+  let fullyStreamedReject;
+  const done = new Promise((resolve, reject) => {
+    fullyStreamedResolve = resolve;
+    fullyStreamedReject = reject;
+  });
+
+  if (process.env.NODE_ENV !== 'production') {
+    logger.log(`About to concatenate ${readerPromises.length} sources ` +
+      `into a final ReadableStream.`);
+  }
+
+  let i = 0;
+  let logStartOfGroup = true;
+
+  const stream = new ReadableStream({
+    pull(controller) {
+      return readerPromises[i]
+        .then((reader) => {
+          if (process.env.NODE_ENV !== 'production') {
+            if (logStartOfGroup) {
+              logStartOfGroup = false;
+              logger.groupCollapsed('Current source is', sourcePromises[i]);
+            }
+          }
+          return reader.read();
+        }).then((result) => {
+          if (result.done) {
+            logStartOfGroup = true;
+            if (process.env.NODE_ENV !== 'production') {
+              logger.log('Reached the end of that source.');
+              logger.groupEnd();
```

Comment:
Dynamic log groups like this will cause issues - we had with some of the previous strategies. Instead you need to stash the logs you want to make and then log everything together.

Reply:
Thanks for the heads-up about nesting and groups. I've refactored that.

I don't have access to the new `Response` at this point in the code, only to the `ReadableStream`. That doesn't end up exposing any useful information when it's logged to the console.

I'm going to take advantage of the fact that the `Response` will normally be logged by folks who use our `Router` class, and consider that sufficient.

Reply:
It now looks like:
<img width="1076" alt="screen shot 2018-04-23 at 2 07 06 pm" src="https://user-images.githubusercontent.com/1749548/39144726-aa5d91f4-46ff-11e8-9536-d9e4635d5929.png">